internal/test/e2e/e2e_test.go

```diff
@@ -162,88 +186,112 @@ func TestE2E_Kind(t *testing.T) {
 		return nil
 	}, time.Minute, time.Second*5)
 	require.NoError(t, err)
-
-	t.Log("Running tests")
-	runAllTests(t, kubectl)
 }
 
-func TestE2E_CurrentContext(t *testing.T) {
-	if !*fUseCurrentContext {
-		t.Skip()
-	}
-
-	home, err := os.UserHomeDir()
-	require.NoError(t, err)
-	configPath := filepath.Join(home, ".kube", "config")
-	if path, found := os.LookupEnv("KUBECONFIG"); found {
-		configPath = path
-	}
-
-	kubectl := &kubectlContext{
-		t:          t,
-		configPath: configPath,
-	}
-
-	runAllTests(t, kubectl)
-}
-
-// Starts a Kind cluster on the local machine, exposing port 2379 accepting ETCD connections.
-func setupLocalCluster(t *testing.T) (*cluster.Context, func()) {
-	t.Log("Starting Kind cluster")
-	kind := cluster.NewContext("etcd-e2e")
-
-	err := kind.Create(create.WithV1Alpha3(&kindv1alpha3.Cluster{
-		TypeMeta: metav1.TypeMeta{
-			Kind:       "Cluster",
-			APIVersion: "kind.sigs.k8s.io/v1alpha3",
-		},
-		Nodes: []kindv1alpha3.Node{
-			{
-				Role: "control-plane",
-				ExtraPortMappings: []cri.PortMapping{
-					{
-						ContainerPort: 32379,
-						HostPort:      2379,
-					},
-				},
-			},
-		},
-	}))
-	require.NoError(t, err)
-
-	tearDown := func() {
-		if !*fCleanup {
-			return
+func TestE2E(t *testing.T) {
+	var (
+		kubectl *kubectlContext
+	)
+	switch {
+	case *fUseKind:
+		var (
+			kind     *cluster.Context
+			stopKind func() error
+			imageTar string
+			wg       sync.WaitGroup
+		)
+		defer func() {
+			if stopKind != nil {
+				err := stopKind()
+				require.NoError(t, err)
+			}
+		}()
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			var err error
+			imageTar, err = buildOperator(ctx)
+			if err != nil {
+				cancel()
+				require.NoError(t, err)
+			}
+		}()
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			var err error
+			// Create Kind cluster to run the workloads.
+			kind, stopKind, err = setupLocalCluster(ctx)
+			if err != nil {
+				cancel()
+				require.NoError(t, err)
+			}
+		}()
+		wg.Wait()
+		require.NoError(t, ctx.Err())
+
+		// Ensure Kind gets stopped on SIGINT/SIGTERM.
+		sigs := make(chan os.Signal)
+		signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+		go func() {
+			<-sigs
+			err := stopKind()
+			require.NoError(t, err)
+		}()
```

Comment:
for safety, do you think it's possible do this earlier? in a similar way to the deferred call that will only call the function if it's been set